### ISSUE TITLE
How can I trace a non-exported function / determine dylib base addr on iOS?

### ISSUE THREAD

**Issue by samyk** [frida/frida#541]
Hi! Frida is **awesome**!

tl;dr *how can I determine a dylib based address or function address to use `frida-trace -a` on iOS?*

I'm investigating why [SSL Kill Switch2](https://github.com/nabla-c0d3/ssl-kill-switch2) (SKS) is not properly allowing TLS cert MITMA's on iOS 11.1.2 via Electra (it is getting hooked into processes but errors on libboringssl.dylib's `boringssl_session_finish_handshake`).

I'm using Frida to investigate this further, so I've created a simple test app using `NSURLConnection` ([https://samy.pl/o/nsurl.tgz](https://samy.pl/o/nsurl.tgz)). What's interesting is that whether or not I inject SKS, Frida never displays/produces a handler for `boringssl_session_finish_handshake`.

I've extracted libboringssl.dylib from my device's dyld_shared_cache and in IDA, I see the function but see it's not exported so I suspect that's why `frida-trace` doesn't pick it up:
![https://samy.pl/dd/bssl.png](https://samy.pl/dd/bssl.png)

I know for a fact it's being executed because SKS fails here:
`Jul 1 21:40:07 the-titanic nsurl-request(libboringssl.dylib)[897] <Error>: Function boringssl_session_finish_handshake: line 2643 The peer was not authenticated. Disconnecting the connection.`
and that string is referenced only in `boringssl_session_finish_handshake`:
![https://samy.pl/dd/bssl2.png](https://samy.pl/dd/bssl2.png)

Trying `0x181C26888` as the address offset doesn't seem to be correct:
```sh
tigerblood:/Users/samy/Code/ios/ssl-kill-switch2$ frida-trace -U nsurl-request -a 'libboringssl.dylib!0x181C26888'
Instrumenting functions...
sub_181c26888: Auto-generated handler at "/Users/samy/Code/ios/ssl-kill-switch2/__handlers__/libboringssl.dylib/sub_181c26888.js"
Error: Skipping 'sub_181c26888': access violation accessing 0x309a74888
Started tracing 1 function. Press Ctrl+C to stop.
```

Running frida-trace does not show the function as we see here:
```sh
tigerblood:/Users/samy/Code/ios/ssl-kill-switch2$ frida-trace -U -i "*boringssl_session_*" nsurl-request
Instrumenting functions...
boringssl_session_get_state: Loaded handler at "/Users/samy/Code/ios/ssl-kill-switch2/__handlers__/libboringssl.dylib/boringssl_session_get_state.js"
boringssl_session_set_bio: Loaded handler at "/Users/samy/Code/ios/ssl-kill-switch2/__handlers__/libboringssl.dylib/boringssl_session_set_bio.js"
boringssl_session_disconnect: Loaded handler at "/Users/samy/Code/ios/ssl-kill-switch2/__handlers__/libboringssl.dylib/boringssl_session_disconnect.js"
boringssl_session_destroy: Loaded handler at "/Users/samy/Code/ios/ssl-kill-switch2/__handlers__/libboringssl.dylib/boringssl_session_destroy.js"
boringssl_session_create: Loaded handler at "/Users/samy/Code/ios/ssl-kill-switch2/__handlers__/libboringssl.dylib/boringssl_session_create.js"
boringssl_session_handshake_start: Loaded handler at "/Users/samy/Code/ios/ssl-kill-switch2/__handlers__/libboringssl.dylib/boringssl_session_handshake_start.js"
boringssl_session_handshake_continue: Loaded handler at "/Users/samy/Code/ios/ssl-kill-switch2/__handlers__/libboringssl.dylib/boringssl_session_handshake_continue.js"
boringssl_session_read: Loaded handler at "/Users/samy/Code/ios/ssl-kill-switch2/__handlers__/libboringssl.dylib/boringssl_session_read.js"
boringssl_session_cache_get_locked_instance: Loaded handler at "/Users/samy/Code/ios/ssl-kill-switch2/__handlers__/libboringssl.dylib/boringssl_session_cache_get_lock_ee2383aa.js"
boringssl_session_write: Loaded handler at "/Users/samy/Code/ios/ssl-kill-switch2/__handlers__/libboringssl.dylib/boringssl_session_write.js"
Started tracing 10 functions. Press Ctrl+C to stop.
           /* TID 0x1203 */
  2958 ms  boringssl_session_read()
  2960 ms  boringssl_session_disconnect()
  2962 ms  boringssl_session_destroy()
           /* TID 0x50f */
436326 ms  boringssl_session_create()
436327 ms  boringssl_session_set_bio()
436327 ms  boringssl_session_cache_get_locked_instance()
436327 ms  boringssl_session_handshake_start()
436327 ms  boringssl_session_get_state()
436328 ms  boringssl_session_handshake_continue()
436328 ms  boringssl_session_get_state()
436328 ms  boringssl_session_handshake_continue()
436328 ms  boringssl_session_get_state()
436360 ms  boringssl_session_handshake_continue()
436361 ms  boringssl_session_get_state()
436361 ms  boringssl_session_handshake_continue()
436361 ms  boringssl_session_get_state()
436362 ms  boringssl_session_handshake_continue()
436362 ms  boringssl_session_get_state()
436362 ms  boringssl_session_handshake_continue()
436362 ms  boringssl_session_get_state()
436400 ms  boringssl_session_handshake_continue()
436401 ms  boringssl_session_get_state()
436401 ms  boringssl_session_handshake_continue()
436401 ms  boringssl_session_get_state()
436439 ms  boringssl_session_handshake_continue()
436439 ms     | boringssl_session_cache_get_locked_instance()
436440 ms  boringssl_session_get_state()
436443 ms  boringssl_session_get_state()
436444 ms  boringssl_session_read()
436444 ms  boringssl_session_read()
436444 ms  boringssl_session_write()
436474 ms  boringssl_session_read()
436474 ms  boringssl_session_read()
436474 ms  boringssl_session_read()
436475 ms  boringssl_session_read()
           /* TID 0x4353 */
436559 ms  boringssl_session_write()
           /* TID 0x50f */
436629 ms  boringssl_session_read()
436630 ms  boringssl_session_read()
436630 ms  boringssl_session_read()
436630 ms  boringssl_session_read()
```

Any thoughts or suggestions on further investigating? Thanks!

**Comment by mrmacete** [comment]
well, to my understading you can get the offset in this way using frida CLI:

```
$ frida -U -f com.supermorte.nsurl-request
     ____
    / _  |   Frida 11.0.12 - A world-class dynamic instrumentation toolkit
   | (_| |
    > _  |   Commands:
   /_/ |_|       help      -> Displays the help system
   . . . .       object?   -> Display information about 'object'
   . . . .       exit/quit -> Exit
   . . . .
   . . . .   More info at http://www.frida.re/docs/home/
Spawned `com.supermorte.nsurl-request`. Use %resume to let the main thread start executing!
[iPhone::com.supermorte.nsurl-request]-> %resume
[iPhone::com.supermorte.nsurl-request]-> DebugSymbol.fromName('boringssl_session_finish_handshake').address.sub(Module.findBaseAddress('libboringssl.dylib'))
"0x14888"
```

and then use that in `frida-trace -a`:

```
$ frida-trace -U -f com.supermorte.nsurl-request -a 'libboringssl.dylib!0x14888'
```

but probably it's simpler to use `frida-trace -s`:

```
frida-trace -U -f com.supermorte.nsurl-request -s 'boringssl_session_finish_handshake'
```

note: i didn't try this with ssl killer tweak, but frida-trace seems happy

**Comment by samyk** [comment]
Thank you, both of those methods worked!

**Comment by antoniozekic** [comment]
@samyk Have you managed to solve this?

I'm trying to re-implement ssl-kill-switch2 functionality in Frida that would work with iOS 10.x and 11.x but without any success. Please check [#687](https://github.com/frida/frida/issues/687) and let me know what you think?

**Comment by antoniozekic** [comment]
Well, I'm reversing CFNetwork, libboringssl and libnetwork so I'll hopefully have some result soon, if not with *create_peer_trust then I'll find some other approach and working bypass, but I would also love to understand why this hook doesn't work... Also, I can't find anyone to confirm on which version did it work and find `dyld_shared_cache_arm64` from that exact iOS version to see what changed...
